prow/github/client.go

```diff
@@ -829,40 +832,45 @@ func (c *Client) GetRepos(org string, isUser bool) ([]Repo, error) {
 	} else {
 		nextURL = fmt.Sprintf("/orgs/%s/repos", org)
 	}
-	for nextURL != "" {
-		resp, err := c.requestRetry(http.MethodGet, nextURL, "", nil)
-		if err != nil {
-			return nil, err
-		}
-		defer resp.Body.Close()
-		if resp.StatusCode < 200 || resp.StatusCode > 299 {
-			return nil, fmt.Errorf("return code not 2XX: %s", resp.Status)
-		}
-
-		b, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			return nil, err
-		}
-
-		var reps []Repo
-		if err := json.Unmarshal(b, &reps); err != nil {
-			return nil, err
-		}
-		repos = append(repos, reps...)
-		nextURL = parseLinks(resp.Header.Get("Link"))["next"]
+	err := c.readPaginatedResults(
+		nextURL, // path
+		"",      // accept
+		func() interface{} { // newObj
+			return &[]Repo{}
+		},
+		func(obj interface{}) { // accumulate
+			repos = append(repos, *(obj.(*[]Repo))...)
+		},
+	)
+	if err != nil {
+		return nil, err
 	}
 	return repos, nil
 }
 
-func (c *Client) GetBranches(org, repo string) ([]Branch, error) {
+// GetBranches returns all repos in an org.
+//
+// If onlyProtected is true it will only return repos with protection enabled,
+// and branch.Protected will be true.
+//
+// This call uses multiple API tokens when results are paginated.
+func (c *Client) GetBranches(org, repo string, onlyProtected bool) ([]Branch, error) {
 	c.log("GetBranches", org, repo)
 	var branches []Branch
-	_, err := c.request(&request{
-		method:    http.MethodGet,
-		path:      fmt.Sprintf("/repos/%s/%s/branches", org, repo),
-		exitCodes: []int{200},
-	}, &branches)
-	return branches, err
+	err := c.readPaginatedResults(
+		fmt.Sprintf("/repos/%s/%s/branches?protected=%t", org, repo, onlyProtected),
```

Comment:
Does this work properly for error states?

https://api.github.com/repos/kubernetes/kubernetes/branches?protected=true

outputs

```
{
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3/repos/branches/#list-branches"
}
```

Reply:
That error happens for authenticated requests too.

I don't think  readPaginatedResults does what you think:

    func (c *Client) readPaginatedResults(path, accept string, newObj func() interface{}, accumulate func(interface{})) error {
        url := fmt.Sprintf("%s%s?per_page=100", c.base, path)

This will fetch `https://api.github.com/repos/kubernetes/kubernetes/branches?protected=true?per_page=100`, and GitHub ignores the first ?protected=true.

Reply:
Done

Thanks for noting the query was bad!

* `?` is a valid value in the query string so we wind up setting the value of protected to `true?per_page=100`, which obviously evaluates to false (or is dropped)
* `https://api.github.com/repos/kubernetes/kubernetes/branches?protected=true` works with the right creds (probably need to be a repo admin)